src/engine/sparqlExpressions/StdevExpression.h

```diff
@@ -0,0 +1,161 @@
+#include "engine/sparqlExpressions/LiteralExpression.h"
+#include "engine/sparqlExpressions/NaryExpression.h"
+#include "engine/sparqlExpressions/SparqlExpression.h"
+#include "engine/sparqlExpressions/SparqlExpressionTypes.h"
+#include "engine/sparqlExpressions/SparqlExpressionValueGetters.h"
+#include "global/ValueId.h"
+
+namespace sparqlExpression {
+
+namespace detail {
+
+/// The STDEV Expression
+
+// Helper function to extract a double from a NumericValue variant
+auto inline numValToDouble =
+    []<typename T>(T&& value) -> std::optional<double> {
+  if constexpr (ad_utility::isSimilar<T, double> ||
+                ad_utility::isSimilar<T, int64_t>) {
+    return static_cast<double>(value);
+  } else {
+    return std::nullopt;
+  }
+};
+
+// Helper expression: The individual deviation squares. A DeviationExpression
+// over X corresponds to the value (X - AVG(X))^2.
+class DeviationExpression : public SparqlExpression {
+ private:
+  Ptr child_;
+
+ public:
+  DeviationExpression(Ptr&& child) : child_{std::move(child)} {}
+
+  // __________________________________________________________________________
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    auto impl =
+        [context](SingleExpressionResult auto&& el) -> ExpressionResult {
+      // Prepare space for result
+      VectorWithMemoryLimit<IdOrLiteralOrIri> exprResult{context->_allocator};
+      std::fill_n(std::back_inserter(exprResult), context->size(),
+                  IdOrLiteralOrIri{Id::makeUndefined()});
+      bool undef = false;
+
+      auto devImpl = [&undef, &exprResult, context](auto generator) {
+        double sum = 0.0;
+        // Intermediate storage of the results returned from the child
+        // expression
+        VectorWithMemoryLimit<double> childResults{context->_allocator};
+
+        // Collect values as doubles
+        for (auto& inp : generator) {
+          const auto& n = detail::NumericValueGetter{}(std::move(inp), context);
+          auto v = std::visit(numValToDouble, n);
+          if (v.has_value()) {
+            childResults.push_back(v.value());
+            sum += v.value();
+          } else {
+            // There is a non-numeric value in the input. Therefore the entire
+            // result will be undef.
+            undef = true;
+            return;
+          }
+          context->cancellationHandle_->throwIfCancelled();
+        }
+
+        // Calculate squared deviation and save for result
+        double avg = sum / static_cast<double>(context->size());
+        for (size_t i = 0; i < childResults.size(); i++) {
+          exprResult.at(i) = IdOrLiteralOrIri{
+              ValueId::makeFromDouble(std::pow(childResults.at(i) - avg, 2))};
+        }
+      };
+
+      auto generator =
+          detail::makeGenerator(AD_FWD(el), context->size(), context);
+      devImpl(std::move(generator));
+
+      if (undef) {
+        return IdOrLiteralOrIri{Id::makeUndefined()};
```

Comment:
You should move as much as possible into a `.cpp`file (e.g. `StdevExpression.cpp`) to make the headers clean and cheap to include.